core/rust/qdbr/src/allocator.rs

```diff
@@ -21,144 +21,488 @@
+    ALLOC_ERROR.with(|error| error.borrow_mut().take())
+}
+
+#[derive(Debug, Copy, Clone)]
+pub enum AllocFailure {
+    /// The underlying allocator failed to allocate memory.
+    OutOfMemory {
+        /// What was requested.
+        requested_size: usize,
+    },
+
+    /// A memory limit would have been exceeded.
+    MemoryLimitExceeded {
+        /// What was requested.
+        requested_size: usize,
+
+        /// The memory tag that was used for the allocation request.
+        memory_tag: i32,
+
+        /// The memory limit at the time of the breach.
+        rss_mem_limit: usize,
+
+        /// The memory used at the time of the breach.
+        rss_mem_used: usize,
+    },
+}
+
+fn save_oom_err(alloc_error: AllocError, requested_size: usize) -> AllocError {
+    ALLOC_ERROR.with(|error| {
+        *error.borrow_mut() = Some(AllocFailure::OutOfMemory { requested_size });
+    });
+    alloc_error
+}
+
+impl Display for AllocFailure {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        match self {
+            AllocFailure::OutOfMemory { requested_size } => {
+                write!(f, "out of memory when allocating {}", requested_size)
+            }
+            AllocFailure::MemoryLimitExceeded {
+                requested_size,
+                memory_tag,
+                rss_mem_limit,
+                rss_mem_used,
+            } => write!(
+                f,
+                "memory limit exceeded when allocating {} with tag {} (rss used: {}, rss limit: {})",
+                requested_size, memory_tag, rss_mem_used, rss_mem_limit
+            ),
+        }
+    }
+}
+
+impl std::error::Error for AllocFailure {}
+
+#[repr(C)]
+pub struct MemTracking {
+    /// Resident set size memory used. Updated on each allocation, reallocation and deallocation,
+    /// also from Java. This is regardless of the memory tag used.
+    rss_mem_used: AtomicUsize,
+
+    /// Resident set size memory limit. Can be updated. Set to 0 for no explicit limit.
+    rss_mem_limit: AtomicUsize,
+
+    /// The total number of allocation calls.
+    malloc_count: AtomicUsize,
+
+    /// The total number of reallocation (grow or shrink) calls.
+    realloc_count: AtomicUsize,
+
+    /// The total number of free calls.
+    free_count: AtomicUsize,
+
+    /// Tracking non-rss memory, such as mmap.
+    _non_rss_mem_used: AtomicUsize,
+}
+
+/// Custom allocator that fails once a memory limit watermark is reached.
+/// It also tracks memory usage for a specific memory tag.
+/// See `Unsafe.java` for the Java side of this.
+#[derive(Clone)]
+#[cfg(not(test))]
+#[repr(C, packed)]
+pub struct QdbAllocator {
+    /// Global counters
+    pub mem_tracking: *const MemTracking,
+
+    /// The total memory used for the specific memory tag.
+    pub tagged_used: *const AtomicUsize,
+
+    /// Memory category. See `MemoryTag.java` for possible values.
+    memory_tag: i32,
+}
+
+#[cfg(test)]
+#[derive(Clone)]
+pub struct QdbAllocator {
```

Comment:
Could you clarify why we have a separate instance for testing?

Reply:
Yeah, this is annoying. It's also the most pragmatic way of doing this.

When running from Java:
* The allocator is created externally.
* The allocators are static for the duration of the program.
* Tests are run sequentially.

When running tests from `cargo test`:
* They are run concurrently by default.
* The allocator can leak past the end of the lifetime of the test because of `rayon`

The last requirement is what means that the allocator type in `#[cfg(test)]` has an `Arc`.